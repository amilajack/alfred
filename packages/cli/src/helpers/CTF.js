import path from 'path';
import fs from 'fs';
import npm from 'npm';
import yarn from 'yarn-api';
import CTF, {
  getDevDependencies,
  CORE_CTFS,
  INTERFACE_STATES,
  normalizeInterfacesOfSkill,
  AddCtfHelpers
} from '@alfredpkg/core';
import type { CtfMap, InterfaceState } from '@alfredpkg/core';
import ValidateConfig from './Validation';
import { getProjectRoot } from './CLI';

export const ENTRYPOINTS = [
  'lib.node.js',
  'app.node.js',
  'lib.browser.js',
  'app.browser.js',
  'lib.electron.main.js',
  'lib.electron.renderer.js',
  'app.electron.main.js',
  'app.electron.renderer.js',
  'lib.react-native.js',
  'app.react-native.js'
];

const projectRoot = getProjectRoot();

export function generateInterfaceStatesFromProject(): Array<InterfaceState> {
  const envs = ['production', 'development', 'test'];
  // Default to development env if no config given
  const env = envs.includes(process.env.NODE_ENV)
    ? process.env.NODE_ENV
    : 'development';

  return ENTRYPOINTS.filter(e =>
    fs.existsSync(path.join(projectRoot, 'src', e))
  ).map(e => {
    const [projectType, target] = e.split('.');
    return {
      env,
      target,
      projectType
    };
  });
}

/**
 * Find all the dependencies that are different between two CTF's.
 * This is used to figure out which deps need to be installed
 */
export function diffCtfDeps(oldCtf: CtfMap, newCtf: CtfMap): Array<string> {
  // Find the dependencies that have changed and install them
  const t: Map<string, string> = new Map();
  const s: Map<string, string> = new Map();

  Object.entries(getDevDependencies(oldCtf)).forEach(([key, val]) => {
    t.set(key, val);
  });
  Object.entries(getDevDependencies(newCtf)).forEach(([key, val]) => {
    if (t.has(key)) {
      if (t.get(key) !== val) {
        throw new Error('Cannot resolve diff deps');
      }
    } else {
      s.set(key, val);
    }
  });

  return Array.from(s.entries()).map(([key, val]) => `${key}@${val}`);
}

/**
 * @TODO Account for `devDependencies` and `dependencies`
 */
export function installDeps(
  dependencies: Array<string> = [],
  npmClient: 'npm' | 'yarn' = 'npm'
): Promise<any> {
  if (!dependencies.length) return Promise.resolve();

  switch (npmClient) {
    case 'npm': {
      return new Promise((resolve, reject) => {
        npm.load({ save: true }, err => {
          if (err) reject(err);

          npm.commands.install(dependencies, (_err, data) => {
            if (_err) reject(_err);
            resolve(data);
          });

          npm.on('log', console.log);
        });
      });
    }
    case 'yarn': {
      return new Promise((resolve, reject) => {
        yarn(['why', 'isobject'], err => {
          if (err) reject(err);
          resolve();
        });
      });
    }
    default: {
      throw new Error('Unsupported npm client. Can only be "npm" or "yarn"');
    }
  }
}

export type AlfredConfig = {
  npmClient: 'npm' | 'yarn',
  skills: Array<string>,
  root: string,
  showConfigs: boolean
};

/**
 * Add skills to a given list of skills to ensure that the list has a complete set
 * of standard ctfs
 */
export function addMissingStdSkillsToCtf(
  ctf: CtfMap,
  alfredConfig: AlfredConfig,
  interfaceState: InterfaceState
): CtfMap {
  const stdCtf = new Map(
    Object.entries({
      lint: CORE_CTFS.eslint,
      format: CORE_CTFS.prettier,
      build: require('@alfredpkg/interface-build').resolveSkill(
        Object.values(CORE_CTFS),
        interfaceState
      ),
      start: require('@alfredpkg/interface-start').resolveSkill(
        Object.values(CORE_CTFS),
        interfaceState
      ),
      test: CORE_CTFS.jest
    })
  );

  ctf.forEach(ctfNode => {
    /* eslint no-param-reassign: off */
    ctfNode.interfaces = normalizeInterfacesOfSkill(ctfNode.interfaces);
  });

  const stdSubCommands: Set<string> = new Set(stdCtf.keys());
  // Create a set of subcommands that the given CTF has
  const ctfSubcommands: Set<string> = Array.from(ctf.values()).reduce(
    (prev, ctfNode) => {
      if (ctfNode.interfaces && ctfNode.interfaces.length) {
        ctfNode.interfaces.forEach(_interface => {
          const { subcommand } = _interface.module;
          prev.add(subcommand);
        });
      }
      return prev;
    },
    new Set()
  );

  stdSubCommands.forEach(stdSubCommand => {
    if (!ctfSubcommands.has(stdSubCommand)) {
      const stdCtfSkillToAdd = stdCtf.get(stdSubCommand);
      if (
        stdCtfSkillToAdd &&
        stdCtfSkillToAdd.interfaces &&
        stdCtfSkillToAdd.interfaces.length
      ) {
        stdCtfSkillToAdd.interfaces = normalizeInterfacesOfSkill(
          stdCtfSkillToAdd.interfaces
        );
      }
      ctf.set(stdCtfSkillToAdd.name, {
        ...stdCtfSkillToAdd,
        ...AddCtfHelpers
      });
    }
  });

  // Add all the CORE_CTF's without subcommands
  ctf.set('babel', { ...CORE_CTFS.babel, ...AddCtfHelpers });
  // @TODO
  // ctf.set('lodash', { ...CORE_CTFS.lodash, ...AddCtfHelpers });

  ctf.forEach(ctfNode => {
    const ctfWithHelpers = {
      ...ctfNode,
      ...AddCtfHelpers
    };
    Object.entries(ctfWithHelpers.ctfs || {}).forEach(([ctfName, ctfFn]) => {
      const correspondingCtfNode = ctf.get(ctfName);
      if (correspondingCtfNode) {
        ctf.set(
          ctfName,
          ctfFn(correspondingCtfNode, ctf, { alfredConfig, ...interfaceState })
        );
      }
    });
  });

  return ctf;
}

export async function loadConfigs(
  pkgPath: string = path.join(projectRoot, 'package.json')
): Promise<{ pkg: Object, pkgPath: string, alfredConfig: AlfredConfig }> {
  if (!fs.existsSync(pkgPath)) {
    throw new Error('Current working directory does not have "package.json"');
  }

  // Read the package.json and validate the Alfred config
  const pkg = JSON.parse((await fs.promises.readFile(pkgPath)).toString());
  const tmpAlfredConfig = pkg.alfred || {};
  ValidateConfig(tmpAlfredConfig || {});

  const defaultOpts = {
    npmClient: 'npm',
    skills: [],
    root: projectRoot
  };
  const alfredConfig = Object.assign({}, defaultOpts, tmpAlfredConfig);

  return { pkg, pkgPath, alfredConfig };
}

export default async function generateCtfFromConfig(
  alfredConfig: AlfredConfig,
  interfaceState: InterfaceState
): Promise<CtfMap> {
  // Check if any valid entrypoints exist
  const states = generateInterfaceStatesFromProject();
  if (!states.length) {
    throw new Error(
      `The project must have at least one entrypoint. Here are some examples of entrypoints:\n\n${ENTRYPOINTS.map(
        e => `"./src/${e}"`
      ).join('\n')}`
    );
  }

  // Generate the CTF
  const tmpCtf: CtfMap = new Map();
  const { skills = [] } = alfredConfig;
  skills.forEach(skill => {
    /* eslint-disable */
    const c = require(skill);
    /* eslint-enable */
    tmpCtf.set(c.name, c);
  });

  const ctf = CTF(Array.from(tmpCtf.values()), alfredConfig, interfaceState);
  addMissingStdSkillsToCtf(ctf, alfredConfig, interfaceState);

  return ctf;
}

export async function diffCtfDepsOfAllInterfaceStates(
  prevAlfredConfig: AlfredConfig,
  currAlfredConfig: AlfredConfig
): Set<string> {
  const stateWithDuplicateDeps = await Promise.all(
    INTERFACE_STATES.map(state =>
      Promise.all([
        generateCtfFromConfig(prevAlfredConfig, state),
        generateCtfFromConfig(currAlfredConfig, state)
      ])
    )
  );

  return Array.from(
    new Set(
      stateWithDuplicateDeps
        .map(([a, b]) => diffCtfDeps(a, b))
        .reduce((prev, curr) => prev.concat(curr), [])
    )
  );
}
