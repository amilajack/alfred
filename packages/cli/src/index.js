// @flow
import path from 'path';
import fs from 'fs';
import { formatPkgJson } from '@alfred/core/lib/config';
import handlebars from 'handlebars';
import type { InterfaceState } from '@alfred/core';

const TEMPLATES_DIR = path.resolve(__dirname, '../templates');

export * from './helpers/parse-input';

async function compileTemplate(templateFilename: string) {
  const source = await fs.promises.readFile(
    path.resolve(TEMPLATES_DIR, templateFilename)
  );
  return handlebars.compile(source.toString(), { noEscape: true });
}

export async function addEntrypoints(
  rawTemplateData: Object,
  root: string,
  entrypointInterfaceStates: Array<InterfaceState>
) {
  const [
    APP_TEMPLATE,
    LIB_TEMPLATE,
    APP_BROWSER_HTML_TEMPLATE,
    TEST_TEMPLATE
  ] = await Promise.all(
    ['app.js.hbs', 'lib.js.hbs', 'index.html.hbs', 'test.js.hbs'].map(
      compileTemplate
    )
  );

  await Promise.all(
    entrypointInterfaceStates.map(interfaceState => {
      const { projectType, target } = interfaceState;
      const isApp = projectType === 'app';
      const isBrowser = target === 'browser';

      const templateData = {
        ...rawTemplateData,
        project: { projectType, target, isApp, isBrowser }
      };

      const writeIndexHtml = () => {
        if (isApp && isBrowser) {
          const content = APP_BROWSER_HTML_TEMPLATE(templateData);
          return fs.promises.writeFile(
            path.join(root, './src/index.html'),
            content
          );
        }
        return Promise.resolve();
      };

      return Promise.all(
        [
          {
            file: `./src/${projectType}.${target}.js`,
            content: (isApp ? APP_TEMPLATE : LIB_TEMPLATE)(templateData)
          },
          {
            file: `./tests/${projectType}.${target}.spec.js`,
            content: TEST_TEMPLATE(templateData)
          }
        ]
          .map(({ file, content }) =>
            fs.promises.writeFile(path.join(root, file), content)
          )
          .concat([writeIndexHtml()])
      );
    })
  );
}

// This function lives here because it is used in both tests and the cli
export async function addBoilerplate(templateData: Object, root: string) {
  const [
    GITIGNORE_TEMPLATE,
    NPM_TEMPLATE,
    README_TEMPLATE,
    EDITORCONFIG_TEMPLATE
  ] = await Promise.all(
    [
      '.gitignore.hbs',
      'package.json.hbs',
      'README.md.hbs',
      '.editorconfig.hbs'
    ].map(compileTemplate)
  );

  await Promise.all(
    [
      {
        file: '.gitignore',
        content: GITIGNORE_TEMPLATE(templateData)
      },
      {
        file: '.editorconfig',
        content: EDITORCONFIG_TEMPLATE(templateData)
      },
      {
        file: 'README.md',
        content: README_TEMPLATE(templateData)
      }
    ].map(({ file, content }) =>
      fs.promises.writeFile(path.join(root, file), content)
    )
  );

  const pkgPath = path.join(root, 'package.json');
  const formattedPkg = await formatPkgJson(
    JSON.parse(NPM_TEMPLATE(templateData))
  );
  await fs.promises.writeFile(pkgPath, formattedPkg);
  const { projectType, target } = templateData.project;

  await addEntrypoints(templateData, root, [
    { projectType, target, env: 'production' }
  ]);
}

export default function getSingleSubcommandFromArgs(
  args: Array<string>
): string {
  switch (args.length) {
    case 0: {
      throw new Error('One subcommand must be passed');
    }
    case 1: {
      break;
    }
    default: {
      throw new Error('Only one subcommand can be passed');
    }
  }

  return args[0];
}
