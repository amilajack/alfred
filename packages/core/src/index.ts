import Project from './project';
import { ProjectInterface } from '@alfred/types';

if (typeof jest !== 'undefined') {
  require('source-map-support/register');
}

export default function alfred(projectDir?: string): Promise<ProjectInterface> {
  const project = new Project(projectDir);
  return project.init();
}

// @TODO @REFACTOR Move this to another module
export { formatPkgJson } from './project';
