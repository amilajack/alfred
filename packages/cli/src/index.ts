import path from 'path';
import fs from 'fs';
import { formatPkgJson } from '@alfred/core';
import handlebars from 'handlebars';
import { Entrypoint, Target } from '@alfred/types';

const TEMPLATES_DIR = path.resolve(__dirname, '../templates');

export type ValidPkgNameResult = {
  validForNewPackages: boolean;
  validForOldPackages: boolean;
  errors?: string[];
  warnings?: string[];
};

export type RawTemplateData = {
  entrypoint?: Entrypoint;
  'alfred-pkg'?: {
    semver: string;
  };
};

export type TemplateData = {
  entrypoint: Entrypoint;
  'alfred-pkg': {
    semver: string;
  };
};

export type GitConfig = {
  user: {
    name: string;
    email: string;
  };
};

async function compileTemplate(
  templateFilename: string
): Promise<HandlebarsTemplateDelegate> {
  const source = await fs.promises.readFile(
    path.resolve(TEMPLATES_DIR, templateFilename)
  );
  return handlebars.compile(source.toString(), { noEscape: true });
}

export async function addEntrypoints(
  rawTemplateData: RawTemplateData,
  root: string,
  targets: Array<Target>
): Promise<void> {
  const [
    APP_TEMPLATE,
    LIB_TEMPLATE,
    APP_BROWSER_HTML_TEMPLATE,
    TEST_TEMPLATE
  ] = await Promise.all(
    ['app.js.hbs', 'lib.js.hbs', 'index.html.hbs', 'test.js.hbs'].map(
      compileTemplate
    )
  );

  await Promise.all(
    targets.map(target => {
      const { project, platform } = target;
      const isApp = project === 'app';
      const isBrowser = platform === 'browser';

      const templateData = {
        ...rawTemplateData,
        project: { project, target, isApp, isBrowser }
      };

      const writeIndexHtml = (): Promise<void> => {
        if (isApp && isBrowser) {
          const content = APP_BROWSER_HTML_TEMPLATE(templateData);
          return fs.promises.writeFile(
            path.join(root, './src/index.html'),
            content
          );
        }
        return Promise.resolve();
      };

      return Promise.all(
        [
          {
            file: `./src/${project}.${target}.js`,
            content: (isApp ? APP_TEMPLATE : LIB_TEMPLATE)(templateData)
          },
          {
            file: `./tests/${project}.${target}.spec.js`,
            content: TEST_TEMPLATE(templateData)
          }
        ]
          .map(({ file, content }) =>
            fs.promises.writeFile(path.join(root, file), content)
          )
          .concat([writeIndexHtml()])
      );
    })
  );
}

// This function lives here because it is used in both tests and the cli
export async function addBoilerplate(
  templateData: TemplateData,
  root: string
): Promise<void> {
  const [
    GITIGNORE_TEMPLATE,
    NPM_TEMPLATE,
    README_TEMPLATE,
    EDITORCONFIG_TEMPLATE
  ] = await Promise.all(
    [
      '.gitignore.hbs',
      'package.json.hbs',
      'README.md.hbs',
      '.editorconfig.hbs'
    ].map(compileTemplate)
  );

  await Promise.all(
    [
      {
        file: '.gitignore',
        content: GITIGNORE_TEMPLATE(templateData)
      },
      {
        file: '.editorconfig',
        content: EDITORCONFIG_TEMPLATE(templateData)
      },
      {
        file: 'README.md',
        content: README_TEMPLATE(templateData)
      }
    ].map(({ file, content }) =>
      fs.promises.writeFile(path.join(root, file), content)
    )
  );

  const pkgPath = path.join(root, 'package.json');
  const formattedPkg = await formatPkgJson(
    JSON.parse(NPM_TEMPLATE(templateData))
  );
  await fs.promises.writeFile(pkgPath, formattedPkg);
  const { project: project, platform: target } = templateData.entrypoint;

  await addEntrypoints(templateData, root, [
    { project: project, platform: target, env: 'production' }
  ]);
}

export function getSingleSubcommandFromArgs(args: Array<string>): string {
  if (args.length === 0) throw new Error('One subcommand must be passed');
  if (args.length > 1) throw new Error('Only one subcommand can be passed');
  return args[0];
}
