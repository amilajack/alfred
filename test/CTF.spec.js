// @flow
/* eslint no-restricted-syntax: off */
import powerset from '@amilajack/powerset';
import * as CtfNodes from '../src/CTF';

const { getConfigs, default: CTF, getDependencies, ...ctfs } = CtfNodes;

describe('CTF', () => {
  const ctfNamesCombinations = powerset(Object.keys(ctfs));

  for (const ctfCombination of ctfNamesCombinations) {
    it(`should test ${ctfCombination.join(',')} ctf combination`, () => {
      expect(ctfCombination).toMatchSnapshot();
      // Get the CTFs for each combination
      const filteredCmfs = ctfCombination.map(ctfName => ctfs[ctfName]);
      expect(getConfigs(CTF(filteredCmfs))).toMatchSnapshot();
      expect(getDependencies(CTF(filteredCmfs))).toMatchSnapshot();
    });
  }
});
