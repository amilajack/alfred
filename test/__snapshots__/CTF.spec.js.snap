// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CTF combination babel 1`] = `
Array [
  "babel",
]
`;

exports[`CTF combination babel 2`] = `
Array [
  Object {
    "extends": "@babel/preset-env",
  },
]
`;

exports[`CTF combination babel 3`] = `
Object {
  "@babel/cli": "7.2.0",
  "@babel/core": "7.2.0",
  "@babel/preset-env": "7.2.0",
}
`;

exports[`CTF combination babel,eslint 1`] = `
Array [
  "babel",
  "eslint",
]
`;

exports[`CTF combination babel,eslint 2`] = `
Array [
  Object {
    "extends": "@babel/preset-env",
  },
  Object {
    "extends": Array [
      "bliss",
    ],
    "parser": "babel-eslint",
  },
]
`;

exports[`CTF combination babel,eslint 3`] = `
Object {
  "@babel/cli": "7.2.0",
  "@babel/core": "7.2.0",
  "@babel/preset-env": "7.2.0",
  "babel-eslint": "5.0.0",
  "eslint": "5.0.0",
}
`;

exports[`CTF combination babel,eslint,react 1`] = `
Array [
  "babel",
  "eslint",
  "react",
]
`;

exports[`CTF combination babel,eslint,react 2`] = `
Array [
  Object {
    "extends": "@babel/preset-env",
    "plugins": Array [
      "@babel/preset-react",
    ],
  },
  Object {
    "extends": Array [
      "bliss",
    ],
    "parser": "babel-eslint",
    "plugins": Array [
      "eslint-plugin-react",
    ],
  },
  "
        import React, { Component } from 'react';

        export default class Root extends Component {
          render() {
            const { store, history } = this.props;
            return (
              <Provider store={store}>
                <ConnectedRouter history={history}>
                  <Routes />
                </ConnectedRouter>
              </Provider>
            );
          }
        }
      ",
  "
        import * as React from 'react';

        export default class App extends React.Component {
          render() {
            const { children } = this.props;
            return <React.Fragment>{children}</React.Fragment>;
          }
        }
      ",
]
`;

exports[`CTF combination babel,eslint,react 3`] = `
Object {
  "@babel/cli": "7.2.0",
  "@babel/core": "7.2.0",
  "@babel/preset-env": "7.2.0",
  "@babel/preset-react": "7.0.0",
  "babel-eslint": "5.0.0",
  "eslint": "5.0.0",
  "eslint-plugin-react": "7.0.0",
  "react": "0.16.0",
}
`;

exports[`CTF combination babel,eslint,webpack 1`] = `
Array [
  "babel",
  "eslint",
  "webpack",
]
`;

exports[`CTF combination babel,eslint,webpack 2`] = `
Array [
  Object {
    "extends": "@babel/preset-env",
  },
  Object {
    "extends": Array [
      "bliss",
    ],
    "parser": "babel-eslint",
    "settings": Object {
      "import/resolver": Object {
        "webpack": Object {
          "config": "configs/webpack.config.js",
        },
      },
    },
  },
  Object {
    "module": Object {
      "devtool": "source-map",
      "entry": "./app/main.dev",
      "mode": "production",
      "output": Object {
        "filename": "./app/main.prod.js",
        "path": "app",
      },
      "rules": Array [
        Object {
          "exclude": /node_modules/,
          "test": /\\\\\\.jsx\\?\\$/,
          "use": Object {
            "loader": "babel-loader",
            "options": Object {
              "cacheDirectory": true,
            },
          },
        },
      ],
      "target": "electron-main",
    },
    "output": Object {
      "libraryTarget": "commonjs2",
      "path": "path.join(__dirname, '..', 'app')",
    },
    "plugins": Array [],
    "resolve": Object {
      "extensions": Array [
        ".js",
        ".json",
      ],
    },
  },
]
`;

exports[`CTF combination babel,eslint,webpack 3`] = `
Object {
  "@babel/cli": "7.2.0",
  "@babel/core": "7.2.0",
  "@babel/preset-env": "7.2.0",
  "babel-eslint": "5.0.0",
  "babel-loader": "5.0.0",
  "eslint": "5.0.0",
  "webpack": "5.0.0",
}
`;

exports[`CTF combination babel,eslint,webpack,react 1`] = `
Array [
  "babel",
  "eslint",
  "webpack",
  "react",
]
`;

exports[`CTF combination babel,eslint,webpack,react 2`] = `
Array [
  Object {
    "extends": "@babel/preset-env",
    "plugins": Array [
      "@babel/preset-react",
    ],
  },
  Object {
    "extends": Array [
      "bliss",
    ],
    "parser": "babel-eslint",
    "plugins": Array [
      "eslint-plugin-react",
    ],
    "settings": Object {
      "import/resolver": Object {
        "webpack": Object {
          "config": "configs/webpack.config.js",
        },
      },
    },
  },
  Object {
    "module": Object {
      "devtool": "source-map",
      "entry": "./app/main.dev",
      "mode": "production",
      "output": Object {
        "filename": "./app/main.prod.js",
        "path": "app",
      },
      "rules": Array [
        Object {
          "exclude": /node_modules/,
          "test": /\\\\\\.jsx\\?\\$/,
          "use": Object {
            "loader": "babel-loader",
            "options": Object {
              "cacheDirectory": true,
            },
          },
        },
      ],
      "target": "electron-main",
    },
    "output": Object {
      "libraryTarget": "commonjs2",
      "path": "path.join(__dirname, '..', 'app')",
    },
    "plugins": Array [],
    "resolve": Object {
      "extensions": Array [
        ".jsx",
        ".json",
      ],
    },
  },
  "
        import React, { Component } from 'react';

        export default class Root extends Component {
          render() {
            const { store, history } = this.props;
            return (
              <Provider store={store}>
                <ConnectedRouter history={history}>
                  <Routes />
                </ConnectedRouter>
              </Provider>
            );
          }
        }
      ",
  "
        import * as React from 'react';

        export default class App extends React.Component {
          render() {
            const { children } = this.props;
            return <React.Fragment>{children}</React.Fragment>;
          }
        }
      ",
]
`;

exports[`CTF combination babel,eslint,webpack,react 3`] = `
Object {
  "@babel/cli": "7.2.0",
  "@babel/core": "7.2.0",
  "@babel/preset-env": "7.2.0",
  "@babel/preset-react": "7.0.0",
  "babel-eslint": "5.0.0",
  "babel-loader": "5.0.0",
  "eslint": "5.0.0",
  "eslint-plugin-react": "7.0.0",
  "react": "0.16.0",
  "webpack": "5.0.0",
}
`;

exports[`CTF combination babel,react 1`] = `
Array [
  "babel",
  "react",
]
`;

exports[`CTF combination babel,react 2`] = `
Array [
  Object {
    "extends": "@babel/preset-env",
    "plugins": Array [
      "@babel/preset-react",
    ],
  },
  "
        import React, { Component } from 'react';

        export default class Root extends Component {
          render() {
            const { store, history } = this.props;
            return (
              <Provider store={store}>
                <ConnectedRouter history={history}>
                  <Routes />
                </ConnectedRouter>
              </Provider>
            );
          }
        }
      ",
  "
        import * as React from 'react';

        export default class App extends React.Component {
          render() {
            const { children } = this.props;
            return <React.Fragment>{children}</React.Fragment>;
          }
        }
      ",
]
`;

exports[`CTF combination babel,react 3`] = `
Object {
  "@babel/cli": "7.2.0",
  "@babel/core": "7.2.0",
  "@babel/preset-env": "7.2.0",
  "@babel/preset-react": "7.0.0",
  "react": "0.16.0",
}
`;

exports[`CTF combination babel,webpack 1`] = `
Array [
  "babel",
  "webpack",
]
`;

exports[`CTF combination babel,webpack 2`] = `
Array [
  Object {
    "extends": "@babel/preset-env",
  },
  Object {
    "module": Object {
      "devtool": "source-map",
      "entry": "./app/main.dev",
      "mode": "production",
      "output": Object {
        "filename": "./app/main.prod.js",
        "path": "app",
      },
      "rules": Array [
        Object {
          "exclude": /node_modules/,
          "test": /\\\\\\.jsx\\?\\$/,
          "use": Object {
            "loader": "babel-loader",
            "options": Object {
              "cacheDirectory": true,
            },
          },
        },
      ],
      "target": "electron-main",
    },
    "output": Object {
      "libraryTarget": "commonjs2",
      "path": "path.join(__dirname, '..', 'app')",
    },
    "plugins": Array [],
    "resolve": Object {
      "extensions": Array [
        ".js",
        ".json",
      ],
    },
  },
]
`;

exports[`CTF combination babel,webpack 3`] = `
Object {
  "@babel/cli": "7.2.0",
  "@babel/core": "7.2.0",
  "@babel/preset-env": "7.2.0",
  "babel-loader": "5.0.0",
  "webpack": "5.0.0",
}
`;

exports[`CTF combination babel,webpack,react 1`] = `
Array [
  "babel",
  "webpack",
  "react",
]
`;

exports[`CTF combination babel,webpack,react 2`] = `
Array [
  Object {
    "extends": "@babel/preset-env",
    "plugins": Array [
      "@babel/preset-react",
    ],
  },
  Object {
    "module": Object {
      "devtool": "source-map",
      "entry": "./app/main.dev",
      "mode": "production",
      "output": Object {
        "filename": "./app/main.prod.js",
        "path": "app",
      },
      "rules": Array [
        Object {
          "exclude": /node_modules/,
          "test": /\\\\\\.jsx\\?\\$/,
          "use": Object {
            "loader": "babel-loader",
            "options": Object {
              "cacheDirectory": true,
            },
          },
        },
      ],
      "target": "electron-main",
    },
    "output": Object {
      "libraryTarget": "commonjs2",
      "path": "path.join(__dirname, '..', 'app')",
    },
    "plugins": Array [],
    "resolve": Object {
      "extensions": Array [
        ".jsx",
        ".json",
      ],
    },
  },
  "
        import React, { Component } from 'react';

        export default class Root extends Component {
          render() {
            const { store, history } = this.props;
            return (
              <Provider store={store}>
                <ConnectedRouter history={history}>
                  <Routes />
                </ConnectedRouter>
              </Provider>
            );
          }
        }
      ",
  "
        import * as React from 'react';

        export default class App extends React.Component {
          render() {
            const { children } = this.props;
            return <React.Fragment>{children}</React.Fragment>;
          }
        }
      ",
]
`;

exports[`CTF combination babel,webpack,react 3`] = `
Object {
  "@babel/cli": "7.2.0",
  "@babel/core": "7.2.0",
  "@babel/preset-env": "7.2.0",
  "@babel/preset-react": "7.0.0",
  "babel-loader": "5.0.0",
  "react": "0.16.0",
  "webpack": "5.0.0",
}
`;

exports[`CTF combination eslint 1`] = `
Array [
  "eslint",
]
`;

exports[`CTF combination eslint 2`] = `
Array [
  Object {
    "extends": Array [
      "bliss",
    ],
  },
]
`;

exports[`CTF combination eslint 3`] = `
Object {
  "eslint": "5.0.0",
}
`;

exports[`CTF combination eslint,react 1`] = `
Array [
  "eslint",
  "react",
]
`;

exports[`CTF combination eslint,react 2`] = `
Array [
  Object {
    "extends": Array [
      "bliss",
    ],
    "plugins": Array [
      "eslint-plugin-react",
    ],
  },
  "
        import React, { Component } from 'react';

        export default class Root extends Component {
          render() {
            const { store, history } = this.props;
            return (
              <Provider store={store}>
                <ConnectedRouter history={history}>
                  <Routes />
                </ConnectedRouter>
              </Provider>
            );
          }
        }
      ",
  "
        import * as React from 'react';

        export default class App extends React.Component {
          render() {
            const { children } = this.props;
            return <React.Fragment>{children}</React.Fragment>;
          }
        }
      ",
]
`;

exports[`CTF combination eslint,react 3`] = `
Object {
  "eslint": "5.0.0",
  "eslint-plugin-react": "7.0.0",
  "react": "0.16.0",
}
`;

exports[`CTF combination eslint,webpack 1`] = `
Array [
  "eslint",
  "webpack",
]
`;

exports[`CTF combination eslint,webpack 2`] = `
Array [
  Object {
    "extends": Array [
      "bliss",
    ],
    "settings": Object {
      "import/resolver": Object {
        "webpack": Object {
          "config": "configs/webpack.config.js",
        },
      },
    },
  },
  Object {
    "module": Object {
      "rules": Array [
        Object {
          "exclude": /node_modules/,
          "test": /\\\\\\.jsx\\?\\$/,
          "use": Object {
            "loader": "babel-loader",
            "options": Object {
              "cacheDirectory": true,
            },
          },
        },
      ],
    },
    "output": Object {
      "libraryTarget": "commonjs2",
      "path": "path.join(__dirname, '..', 'app')",
    },
    "plugins": Array [],
    "resolve": Object {
      "extensions": Array [
        ".js",
        ".json",
      ],
    },
  },
]
`;

exports[`CTF combination eslint,webpack 3`] = `
Object {
  "eslint": "5.0.0",
  "webpack": "5.0.0",
}
`;

exports[`CTF combination eslint,webpack,react 1`] = `
Array [
  "eslint",
  "webpack",
  "react",
]
`;

exports[`CTF combination eslint,webpack,react 2`] = `
Array [
  Object {
    "extends": Array [
      "bliss",
    ],
    "plugins": Array [
      "eslint-plugin-react",
    ],
    "settings": Object {
      "import/resolver": Object {
        "webpack": Object {
          "config": "configs/webpack.config.js",
        },
      },
    },
  },
  Object {
    "module": Object {
      "rules": Array [
        Object {
          "exclude": /node_modules/,
          "test": /\\\\\\.jsx\\?\\$/,
          "use": Object {
            "loader": "babel-loader",
            "options": Object {
              "cacheDirectory": true,
            },
          },
        },
      ],
    },
    "output": Object {
      "libraryTarget": "commonjs2",
      "path": "path.join(__dirname, '..', 'app')",
    },
    "plugins": Array [],
    "resolve": Object {
      "extensions": Array [
        ".jsx",
        ".json",
      ],
    },
  },
  "
        import React, { Component } from 'react';

        export default class Root extends Component {
          render() {
            const { store, history } = this.props;
            return (
              <Provider store={store}>
                <ConnectedRouter history={history}>
                  <Routes />
                </ConnectedRouter>
              </Provider>
            );
          }
        }
      ",
  "
        import * as React from 'react';

        export default class App extends React.Component {
          render() {
            const { children } = this.props;
            return <React.Fragment>{children}</React.Fragment>;
          }
        }
      ",
]
`;

exports[`CTF combination eslint,webpack,react 3`] = `
Object {
  "eslint": "5.0.0",
  "eslint-plugin-react": "7.0.0",
  "react": "0.16.0",
  "webpack": "5.0.0",
}
`;

exports[`CTF combination react 1`] = `
Array [
  "react",
]
`;

exports[`CTF combination react 2`] = `
Array [
  "
        import React, { Component } from 'react';

        export default class Root extends Component {
          render() {
            const { store, history } = this.props;
            return (
              <Provider store={store}>
                <ConnectedRouter history={history}>
                  <Routes />
                </ConnectedRouter>
              </Provider>
            );
          }
        }
      ",
  "
        import * as React from 'react';

        export default class App extends React.Component {
          render() {
            const { children } = this.props;
            return <React.Fragment>{children}</React.Fragment>;
          }
        }
      ",
]
`;

exports[`CTF combination react 3`] = `
Object {
  "react": "0.16.0",
}
`;

exports[`CTF combination webpack 1`] = `
Array [
  "webpack",
]
`;

exports[`CTF combination webpack 2`] = `
Array [
  Object {
    "module": Object {
      "rules": Array [
        Object {
          "exclude": /node_modules/,
          "test": /\\\\\\.jsx\\?\\$/,
          "use": Object {
            "loader": "babel-loader",
            "options": Object {
              "cacheDirectory": true,
            },
          },
        },
      ],
    },
    "output": Object {
      "libraryTarget": "commonjs2",
      "path": "path.join(__dirname, '..', 'app')",
    },
    "plugins": Array [],
    "resolve": Object {
      "extensions": Array [
        ".js",
        ".json",
      ],
    },
  },
]
`;

exports[`CTF combination webpack 3`] = `
Object {
  "webpack": "5.0.0",
}
`;

exports[`CTF combination webpack,react 1`] = `
Array [
  "webpack",
  "react",
]
`;

exports[`CTF combination webpack,react 2`] = `
Array [
  Object {
    "module": Object {
      "rules": Array [
        Object {
          "exclude": /node_modules/,
          "test": /\\\\\\.jsx\\?\\$/,
          "use": Object {
            "loader": "babel-loader",
            "options": Object {
              "cacheDirectory": true,
            },
          },
        },
      ],
    },
    "output": Object {
      "libraryTarget": "commonjs2",
      "path": "path.join(__dirname, '..', 'app')",
    },
    "plugins": Array [],
    "resolve": Object {
      "extensions": Array [
        ".jsx",
        ".json",
      ],
    },
  },
  "
        import React, { Component } from 'react';

        export default class Root extends Component {
          render() {
            const { store, history } = this.props;
            return (
              <Provider store={store}>
                <ConnectedRouter history={history}>
                  <Routes />
                </ConnectedRouter>
              </Provider>
            );
          }
        }
      ",
  "
        import * as React from 'react';

        export default class App extends React.Component {
          render() {
            const { children } = this.props;
            return <React.Fragment>{children}</React.Fragment>;
          }
        }
      ",
]
`;

exports[`CTF combination webpack,react 3`] = `
Object {
  "react": "0.16.0",
  "webpack": "5.0.0",
}
`;
